Add `dynamic_bitset`

amukkara opened the pull request:
This PR adds `dynamic_bitset` code that will be used in Trie data structure. Trie will be integrated in a separate PR #350.

Since `dynamic_bitset` is not intended to be part of public-facing API, all files (.cuh and .inl) are located in include/cuco/detail/trie/dynamic_bitset

Tests are added in tests/dynamic_bitset

amukkara committed:
Initial checkin of bitvector

Added simple size test

amukkara committed:
Add bitvector get test

amukkara committed:
Add more bitvector tests

amukkara committed:
Add trie tests

amukkara committed:
Move implementations to .inl files

amukkara committed:
Add bit_vector reference classes

amukkara committed:
Use aow_storage for bit_vector structures

amukkara committed:
Minor

amukkara committed:
Remove trie

amukkara committed:
Clang format

amukkara committed:
Coalesce all bitvector operations into single tag

amukkara committed:
Naming

amukkara committed:
Allocate and size aow arrays after host side build

Possible by shifting from static to dynamic allocation

amukkara committed:
Handle empty bitvector with zero bits

amukkara committed:
Minor

amukkara committed:
Remove bitvector memory footprint logic

amukkara committed:
Doxygen comments

amukkara committed:
Avoid C-style casts

amukkara committed:
Consistent use of 64-bit integer types

amukkara committed:
Rename method from `add` to `append`

amukkara committed:
Remove `Key` template parameter

This was only used in one method and
is replaced by size_type

amukkara committed:
Minor

amukkara committed:
Cherrypick bitvector files from trie branch

GPUtester commented:
Can one of the admins verify this patch?

Admins can comment `ok to test` to allow this one PR to run or `add to allowlist` to allow all future PRs from the same author to run.

PointKernel commented:
add to allowlist

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+struct rank {
+  uint32_t abs_hi_;              ///< Upper 32 bits of base
+  uint8_t abs_lo_;               ///< Lower 8 bits of base
+  std::array<uint8_t, 3> rels_;  ///< Offsets for 64-bit sub-intervals
```
```suggestion
  cuda::std::array<uint8_t, 3> rels_;  ///< Offsets for 64-bit sub-intervals
```
`cuda::std::array` is preferred since the data would be accessed from device

amukkara replied:
3894d5a984023a9e417c7f9e409f742d9b76fc76

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+   */
+  void set_abs(uint64_t abs)
+  {
+    abs_hi_ = static_cast<uint32_t>(abs >> 8);
```
Note to myself. Should find a way to get rid of magic numbers like `8` and `3`

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+   *
+   * @return The base rank
+   */
+  __host__ __device__ uint64_t abs() const
```
```suggestion
  __host__ __device__ constexpr uint64_t abs() const noexcept
```
`constexpr` and `noexcept` when possible

amukkara replied:
10c09006eea55817ad931214a4b40685611a6fef

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+   *
+   * @param abs Base rank
+   */
+  void set_abs(uint64_t abs)
```
```suggestion
  constexpr void set_abs(uint64_t abs) noexcept
```

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+template <class Extent             = cuco::experimental::extent<std::size_t>,
+          cuda::thread_scope Scope = cuda::thread_scope_device,
+          class Allocator          = cuco::cuda_allocator<std::byte>,
+          class Storage            = cuco::experimental::storage<1>>
```
```suggestion
template <class Allocator          = cuco::cuda_allocator<std::byte>>
```

- `Scope` seems not related to this work
- Not sure bit vector would benefit from the cuco extent type, we could just use `using size_type = size_t;` internally
- Storage doesn't need to be exposed since tuning `window_size` doesn't apply in this case.

amukkara replied:
d063cc50bc0b5ad0b02b8f032ba75c4a55b2f968

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+   * @param index position of bit to be modified
+   * @param bit new value of bit
+   */
+  void set(size_type index, bool bit);
```
```suggestion
  void set(size_type index, bool bit) noexcept;
```
Similar to other cuco host operations: would a "bulk" host API work here? Something like:
```cpp
void set(std::span<size_type> indices, std::span<bool> bit) noexcept;
```

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+   *
+   * @param bit Boolean value of new bit to be added
+   */
+  void append(bool bit);
```
```suggestion
  void append(bool bit) noexcept;
```

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+   *
+   * Also creates device-side snapshot
+   */
+  void build();
```
```suggestion
  void build() noexcept;
```

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+   *
+   * @param bit new value of last bit
+   */
+  void set_last(bool bit);
```
```suggestion
  void set_last(bool bit) noexcept;
```

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+   */
+  void build();
+
+  using size_type = typename Extent::value_type;  ///< size type to specify bit index
```
```suggestion
  using size_type = std::size_t;  ///< size type to specify bit index
```
And use `size_type` throughout the whole class.

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+  static constexpr auto thread_scope = Scope;  ///< CUDA thread scope
+
+  using extent_type =
+    decltype(make_window_extent<cg_size, window_size>(std::declval<Extent>()));  ///< Extent type
```
```suggestion
```

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+  using extent_type =
+    decltype(make_window_extent<cg_size, window_size>(std::declval<Extent>()));  ///< Extent type
+  using allocator_type = Allocator;                                              ///< Allocator type
+  using storage_type =
+    detail::storage<Storage, size_type, extent_type, allocator_type>;  ///< Storage type
```
```suggestion
  using slot_type = uint64_t;
  using storage_type =
    aow_storage<slot_type, 1, size_type, allocator_type>;  ///< Storage type
```
The slot type should be independent of `size_type`. Imagine `using size_type = int32_t`;
Also, can use `aow_storage` directly to simplify the logic.

amukkara replied:
d063cc50bc0b5ad0b02b8f032ba75c4a55b2f968

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+          class Storage            = cuco::experimental::storage<1>>
+class bit_vector {
+ public:
+  bit_vector();
```
```suggestion
  bit_vector(Allocator const& allocator = Allocator{});
```
We need to at least pass a default allocator

amukkara replied:
a371e27b6c612e7a61bee2fe6a6e70654f006051

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+namespace experimental {
+
+/**
+ * @brief Struct to store ranks of bits at 256-bit intervals
```
Is 256 chosen for performance purposes?

amukkara replied:
It is a tradeoff between space efficiency and performance. Larger values will create a smaller index, but searches will be slower.

This is hardcoded to 256 (four 64-bit words) now, but can be made configurable. Main change would be in `rank` structure which stores a single base and deltas for each 64-bit word.

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+  size_t size() const { return n_bits_; }
+
+ private:
+  uint64_t n_bits_;  ///< Number of bits added to bit_vector
```
```suggestion
  size_type n_bits_;  ///< Number of bits added to bit_vector
```

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,200 @@
+  uint64_t n_bits_;  ///< Number of bits added to bit_vector
+
+  // Host-side structures
+  std::vector<uint64_t> words_;     ///< Words vector that represents all bits
+  std::vector<rank> ranks_;         ///< Holds the rank values for every 256-th bit (4-th word)
+  std::vector<rank> ranks0_;        ///< Same as ranks_ but for `0` bits
+  std::vector<uint64_t> selects_;   ///< Holds pointers to (0, 256, 512...)th `1` bit in ranks_
+  std::vector<uint64_t> selects0_;  ///< Same as selects_, but for `0` bits
```
```suggestion
  std::vector<slot_type> words_;     ///< Words vector that represents all bits
  std::vector<rank> ranks_;         ///< Holds the rank values for every 256-th bit (4-th word)
  std::vector<rank> ranks0_;        ///< Same as ranks_ but for `0` bits
  std::vector<size_type> selects_;   ///< Holds indices to (0, 256, 512...)th `1` bit in ranks_
  std::vector<size_type> selects0_;  ///< Same as selects_, but for `0` bits
```
I could be wrong, but the idea here is to use `size_type` and `slot_type` consistently to avoid plain `uint64_t`.

amukkara replied:
91b628a5925bec0b4a4c7f21dc6ebad40c8ba0d9

PointKernel reviewed (commented):
Thanks for putting up the PR! :fire: 

Just a quick pass of the top-level bitvector header. Some are the issues that need to be solved and some are just notes to myself. Feel free to address reviews on any topics you are comfortable dealing with. @sleeepyjack and I would take care of the code-style or software engineering stuff.

amukkara committed:
Use cuda math instrincs in device functions

amukkara committed:
Use cuda::std::array

Co-authored-by: Yunsong Wang <yunsongw@nvidia.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto code formatting

amukkara committed:
Add constexpr and noexcept specifiers

amukkara committed:
Spacing

amukkara committed:
Remove unnecessary template parameters

amukkara committed:
Allocator argument in constructor

amukkara committed:
Use size_type and slot_type

Avoid using plain uint64_t

amukkara committed:
Explicitly define frequently used constants

amukkara committed:
Comments

amukkara committed:
Improve order of function implementations

For better readability

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,223 @@
+
+  const size_type bits_per_word   = sizeof(slot_type) * 8;  ///< Bits in a word
+  const size_type words_per_block = 4;  ///< Provides tradeoff between space efficiency and perf.
+  const size_type bits_per_block  = words_per_block * bits_per_word;
```
```suggestion
  constexpr size_type bits_per_word   = sizeof(slot_type) * 8;  ///< Bits in a word
  constexpr size_type words_per_block = 4;  ///< Provides tradeoff between space efficiency and perf.
  constexpr size_type bits_per_block  = words_per_block * bits_per_word;
```
Also, I feel those should be public thus downstream objects can use to eliminate magic numbers

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,223 @@
+}  // namespace cuco
+
+#include <cuco/detail/trie/bit_vector/bit_vector.inl>
+#include <cuco/detail/trie/bit_vector/bit_vector_ref.inl>
```
```suggestion
```

amukkara replied:
4451d7e19a6ee2041a6ae0522877abc474b2d97b

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector_ref.cuh:
```diff
@@ -0,0 +1,47 @@
+};
+
+}  // namespace experimental
+}  // namespace cuco
```
```suggestion
}  // namespace cuco

#include <cuco/detail/trie/bit_vector/bit_vector_ref.inl>
```

amukkara replied:
4451d7e19a6ee2041a6ae0522877abc474b2d97b

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,223 @@
+
+#include <thrust/device_vector.h>
+
+#include <cuda/atomic>
```
```suggestion
```
seems not used

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,223 @@
+   *
+   * @param abs Base rank
+   */
+  void constexpr set_abs(uint64_t abs) noexcept
```
Does this need to be `__host__ __device__`? as `abs()`?

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,223 @@
+ *
+ * @tparam Allocator Type of allocator used for device storage
+ */
+
```
```suggestion
```

amukkara committed:
Refactor selects entry addition

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,223 @@
+#pragma once
+
+#include <cuco/detail/trie/bit_vector/bit_vector_ref.cuh>
+#include <cuco/extent.cuh>
```
```suggestion
```

amukkara replied:
4451d7e19a6ee2041a6ae0522877abc474b2d97b

amukkara committed:
Merge branch 'NVIDIA:dev' into trie-bitvector

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,221 @@
+ * Need this so that all aow_storage structures in bitvector have 64-bit element type
+ */
+union rank_union {
+  uint64_t word_;  ///< word view
```
Note: this should be `slot_type` instead of plain integer.

amukkara replied:
This struct is not needed after c5ec25452ea9a37d355c2eaca787574bea78ec01

amukkara committed:
Bulk bitvector get operation

amukkara committed:
Add device-ref set operation

Test does not pass.
set() does not seem to update bitvector memory.

amukkara committed:
Bulk set API

sleeepyjack commented:
/ok to test

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,238 @@
+ *
+ * @tparam Allocator Type of allocator used for device storage
+ */
+
```
```suggestion
```
Remove empty line so the Doxygen style check won't complain.

amukkara replied:
f74bee9370517e91b1dbcd1af1afcfc9e8052f24

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,238 @@
+template <class Allocator = cuco::cuda_allocator<std::byte>>
+class bit_vector {
+ public:
+  bit_vector(Allocator const& allocator = Allocator{});
```
Missing docs (Doxygen style check will complain)

amukkara replied:
f74bee9370517e91b1dbcd1af1afcfc9e8052f24

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,238 @@
+  /**
+   * @brief adds a new bit at the end
+   *
+   * Grows internal storage if needed
```
I agree with @PointKernel that since this data structure can grow dynamically, we should add a `dynamic_` prefix to the name. We can take care of this.

sleeepyjack replied:
Also, this data structure comes with some additional bells and whistles compared to a plain bit vector and is also specific to the trie implementation. Can we come up with a more descriptive name?

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,238 @@
+
+  const size_type bits_per_word   = sizeof(slot_type) * 8;  ///< Bits in a word
+  const size_type words_per_block = 4;  ///< Provides tradeoff between space efficiency and perf.
+  const size_type bits_per_block  = words_per_block * bits_per_word;
```
```suggestion
  static constexpr size_type bits_per_word   = sizeof(slot_type) * 8;  ///< Bits in a word
  static constexpr size_type words_per_block = 4;  ///< Provides tradeoff between space efficiency and perf.
  static constexpr size_type bits_per_block  = words_per_block * bits_per_word;
```

amukkara replied:
9d5100d2c20347acf18d4d10a30ea76db159f13e

amukkara committed:
Use size_type in tests

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector.inl:
```diff
@@ -0,0 +1,241 @@
+  delete aow_ranks_;
+  delete aow_selects_;
+  delete aow_ranks0_;
+  delete aow_selects0_;
```
It is easy to forget a deletion here, and thus run into a memory leak. One way to ship around that is to define the members as `std::unique_ptr`, which takes care of managing the object lifetime.

@PointKernel would it make sense to apply a PIMPL idiom to this class?

amukkara replied:
Added unique_ptrs eb7f9572fd61fc3d052e53736f4a2b47a7bf72a1

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector.inl:
Since this class is no longer in the public-facing API, it would be ok to move the implementation to the `.cuh` file and remove the `.inl` file. Not a requirement though.

amukkara replied:
I prefer having implementations in .inl to avoid very large files (for readability)

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,238 @@
+  using allocator_type = Allocator;  ///< Allocator type
+  using slot_type      = uint64_t;   ///< Slot type
+  using storage_type =
+    aow_storage<slot_type, 1, extent<size_type>, allocator_type>;  ///< Storage type
```
The advantage of using `aow_storage` is that it enables vectorized loads of windows. Since you're not relying on that feature (window size is 1), I tend to say that we don't need this specialized storage. Instead, you can use `thrust::host_vector` and `thrust::device_vector` for your storage classes, which should also simplify some of the data migration parts.

PointKernel replied:
Using `aow_storage` would also benefit from its member functions like `initialize`. No strong preference here.

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector.inl:
```diff
@@ -0,0 +1,241 @@
+    ranks_(),
+    selects_(),
+    n_bits_(0),
+    allocator_(allocator),
```
Do we actually use this allocator?

PointKernel replied:
This allocator should be passed to all storage data members.

amukkara replied:
allocator is used only once, as an argument to aow_storage constructor.

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector.inl:
```diff
@@ -0,0 +1,241 @@
+
+template <class Allocator>
+bit_vector<Allocator>::bit_vector(Allocator const& allocator)
+  : words_(),
```
Minor style thingy: 
```suggestion
  : words_{},
```

amukkara replied:
e248c9d26f40fcaeca19fa85ce9c313595c6bc03

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector.inl:
```diff
@@ -0,0 +1,241 @@
+
+template <class Allocator>
+template <typename KeyIt, typename OutputIt>
+void bit_vector<Allocator>::get(KeyIt keys_begin,
```
Iterator-based algorithms typically pass the end of the output (i.e., `outputs_end`) as the return value. Not sure if that makes sense here, since there is a 1:1 relationship between inputs and outputs.

PointKernel replied:
This should be fine. Unlike multimap or single map insertion where the output number doesn't necessarily match the input number, we always have one output for a given input

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector.inl:
```diff
@@ -0,0 +1,241 @@
+void bit_vector<Allocator>::build() noexcept
+{
+  build_ranks_and_selects(words_, ranks_, selects_, false);   // 1-bits
+  build_ranks_and_selects(words_, ranks0_, selects0_, true);  // 0-bits
```
This sure would be a candidate for rudimentary parallelization, i.e., one thread per function call.

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector.inl:
```diff
@@ -0,0 +1,241 @@
+                                              std::vector<size_type>& selects) noexcept
+{
+  while (word != 0) {
+    size_type pos = __builtin_ctzll(word);
```
This is a GCC builtin. There's probably a portable way of doing this using libcudacxx backports from the stdlib `<bit>` header.

amukkara replied:
8687daa9ca744bd61826df0a9147f395047a07fd

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector.inl:
```diff
@@ -0,0 +1,241 @@
+                                                   std::vector<T>& host_array) noexcept
+{
+  uint64_t num_elements = host_array.size();
+  *aow                  = new storage_type(extent<size_type>{num_elements + 1}, allocator_);
```
Plain `new` is considered harmful ;)

amukkara replied:
eb7f9572fd61fc3d052e53736f4a2b47a7bf72a1

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector_ref.cuh:
```diff
@@ -0,0 +1,47 @@
+   * @param ranks0_ref Non-owning ref of ranks0 slot storage
+   * @param selects0_ref Non-owning ref of selects0 slot storage
+   */
+  __host__ __device__ explicit constexpr bit_vector_ref(storage_ref_type words_ref,
+                                                        storage_ref_type ranks_ref,
+                                                        storage_ref_type selects_ref,
+                                                        storage_ref_type ranks0_ref,
+                                                        storage_ref_type selects0_ref) noexcept;
+
```
Passing each storage one-by-one is suboptimal, since they might be incompatible (e.g. different lengths). Since this is not user-facing code, it's not a huge problem, but consider wrapping these in a single struct which internally checks for the length requirement.

amukkara replied:
agreed. by using a struct, ranks and words can use different storage types.

we can then delete rank_union struct, whose sole purpose is to be able to use same storage type for words and ranks.

amukkara replied:
relevant changes in c5ec25452ea9a37d355c2eaca787574bea78ec01

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector_ref.cuh:
```diff
@@ -0,0 +1,47 @@
+
+#include <cuco/operator.hpp>
+
+namespace cuco {
+namespace experimental {
```
This is declared in public namespace, but we want it under detail namespace.

amukkara replied:
should all bitvector code go under coco::detail?

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector_ref.cuh:
```diff
@@ -0,0 +1,47 @@
+ private:
+  storage_ref_type words_ref_, ranks_ref_, selects_ref_, ranks0_ref_, selects0_ref_;
+
+  // Mixins need to be friends with this class in order to access private members
+  template <typename Op, typename Ref>
+  friend class detail::operator_impl;
```
I'm reconsidering if the operator mixin design makes sense here. The only device-sided operations are all `const`, so there's no benefit in dispatching optimized code paths based on the operations used. Plain `__device__` member functions should be equivalent and are less complex to implement.

PointKernel replied:
Right, ref dynamic bitset would never "grow" the container itself thus having operator base class doesn't help here. I've moved ref as a nested type of `dynamic_bitset`.

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector_ref.inl:
```diff
@@ -0,0 +1,238 @@
+  using size_type = typename StorageRef::size_type;            ///< Size type
+  using slot_type = typename StorageRef::value_type;           ///< Slot type
+
+  const size_type bits_per_word   = sizeof(slot_type) * 8;
+  const size_type words_per_block = 4;  //< This should match the defintion in bit_vector
```
These can be `constexpr`.

amukkara replied:
9d5100d2c20347acf18d4d10a30ea76db159f13e

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector_ref.inl:
```diff
@@ -0,0 +1,238 @@
+  [[nodiscard]] __device__ size_type find_next_set(size_type key) const noexcept
+  {
+    auto const& ref_  = static_cast<ref_type const&>(*this);
+    size_type word_id = key / bits_per_word;
+    size_type bit_id  = key % bits_per_word;
```
Declaring `bits_per_word`as `constexpr` should speed up these otherwise expensive operations `/` and `%` considerably.

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector_ref.inl:
```diff
@@ -0,0 +1,238 @@
+
+    if (rel_id != 0) { n += rank.rels_[rel_id - 1]; }
+
+    n += __popcll(ref_.words_ref_[word_id][0] & ((1UL << bit_id) - 1));
```
I think we now have `cuda::std::popcount`. Although it's just a style thing, since there's no point in adding portability here.

sleeepyjack commented on tests/CMakeLists.txt:
```diff
@@ -95,3 +95,13 @@ ConfigureTest(STATIC_MULTIMAP_TEST
+###################################################################################################
+# - bit_vector tests ------------------------------------------------------------------------------
+ConfigureTest(BIT_VECTOR_TEST
+    bit_vector/find_next_set_test.cu
+    bit_vector/get_test.cu
+    bit_vector/set_test.cu
+    bit_vector/rank_test.cu
+    bit_vector/select_test.cu
+    bit_vector/size_test.cu)
```
We should move these to `trie/bit_vector/...` in the trie PR

sleeepyjack commented on tests/bit_vector/find_next_set_test.cu:
```diff
@@ -0,0 +1,74 @@
+  }
+}
+
+extern bool modulo_bitgen(uint64_t i);
```
Where is this declared?

amukkara replied:
Added comments to avoid confusion. dbd5313393815583a1a2766880cfda0f45d7bcab

sleeepyjack commented on tests/bit_vector/find_next_set_test.cu:
```diff
@@ -0,0 +1,74 @@
+#include <catch2/catch_test_macros.hpp>
+
+template <class BitVectorRef>
+__global__ void find_next_set_kernel(BitVectorRef ref, size_t n, uint64_t* output)
```
If you add a tparam `ÒutputIt`, you won't need the `thrust::raw_pointer_cast` at the call site.

amukkara replied:
6028a794dff6a1e3daf1a8b4d3980ee7c172601b

sleeepyjack commented on tests/bit_vector/get_test.cu:
```diff
@@ -0,0 +1,70 @@
+  }
+}
+
+bool modulo_bitgen(uint64_t i) { return i % 7 == 0; }
```
Ah, I see. So after linking this function should be available to the other test.

sleeepyjack commented on include/cuco/detail/trie/bit_vector/bit_vector_ref.inl:
```diff
@@ -202,6 +202,37 @@ class operator_impl<op::bv_read_tag, bit_vector_ref<StorageRef, Operators...>> {
+  using ref_type                = bit_vector_ref<StorageRef, Operators...>;  ///< Bitvector ref type
+  using size_type               = typename StorageRef::size_type;            ///< Size type
+  using slot_type               = typename StorageRef::value_type;           ///< Slot type
+  const size_type bits_per_word = sizeof(slot_type) * 8;
```
There's one catch with these mixin operators: You cannot have any runtime members in the operator class because we're casting the operator to the `ref_type`, and after doing that, the operator's members occupy the same space as the `ref_type` members. The `static_cast<ref_type&>(*this)` and then accessing any of `ref_type`'s members will get you in trouble.
In this particular case, you can just declare `bits_per_word` as `constexpr` to get rid of the problem.

amukkara committed:
Add static constexpr

amukkara committed:
Minor coding style

amukkara committed:
Minor

amukkara committed:
cuda::std popcount intrinsic

amukkara committed:
Comments

amukkara committed:
get_word operation

To access a full word in building ranks/selects

amukkara committed:
Comment out set test check

amukkara committed:
Generic template parameters

This avoids thrust::raw_pointer_cast

sleeepyjack commented:
/ok to test

amukkara committed:
Comments

amukkara committed:
Use unique_ptrs

amukkara committed:
Use cuda::std intrinsics

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.inl:
```diff
@@ -0,0 +1,277 @@
+    ranks0_{},
+    selects_{},
+    selects0_{},
+    n_bits_(0),
+    allocator_(allocator),
+    aow_words_(nullptr),
+    aow_ranks_(nullptr),
+    aow_selects_(nullptr),
+    aow_ranks0_(nullptr),
+    aow_selects0_(nullptr)
```
```suggestion
    n_bits_{0},
    allocator_[allocator},
    aow_words_{nullptr},
    aow_ranks_{nullptr},
    aow_selects_{nullptr},
    aow_ranks0_{nullptr},
    aow_selects0_{nullptr}
```
Always prefer curly braces over parentheses in initialization list

amukkara replied:
005bd5d42177a3239e4b23545e9f328e7bb07633

PointKernel reviewed (commented):
Partial review. Some of the comments could be out of date.

amukkara committed:
Curly braces in initialization list

amukkara committed:
Remove unused header

Co-authored-by: Yunsong Wang <yunsongw@nvidia.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto code formatting

amukkara committed:
Merge branch 'trie-bitvector' of https://github.com/amukkara/cuCollections into trie-bitvector

sleeepyjack commented:
/ok to test

amukkara committed:
Collect aow refs in a single struct

rank_union is not needed after this change

amukkara committed:
Fix includes

amukkara committed:
Avoid repeated definition of same magic number

amukkara committed:
Device kernels for rank and select generation

amukkara committed:
Larger bitvector sizes in tests

amukkara committed:
Misc fixes in bitvector build

amukkara committed:
Move constructor

amukkara committed:
Bulk API for rank and select operations

PointKernel commented on include/cuco/detail/trie/bit_vector/bit_vector.cuh:
```diff
@@ -0,0 +1,264 @@
+  // Host-side structures
+  std::vector<slot_type> words_;  ///< Words vector that represents all bits
+
+  // Device-side structures
+  thrust::device_vector<slot_type> d_words_;   ///< Device words vector
+  thrust::device_vector<rank> ranks_;          ///< Rank values for every 256-th bit (4-th word)
+  thrust::device_vector<rank> ranks0_;         ///< Same as ranks_ but for `0` bits
+  thrust::device_vector<size_type> selects_;   ///< Block indices of (0, 256, 512...)th `1` bit
+  thrust::device_vector<size_type> selects0_;  ///< Same as selects_, but for `0` bits
+
+  allocator_type allocator_;  ///< Allocator used to (de)allocate temporary storage
+  std::unique_ptr<words_storage_type> aow_words_;  ///< Array of window storage structure
+  std::unique_ptr<ranks_storage_type> aow_ranks_;
+  std::unique_ptr<ranks_storage_type> aow_ranks0_;
+  std::unique_ptr<selects_storage_type> aow_selects_;
+  std::unique_ptr<selects_storage_type> aow_selects0_;
```
Maybe a leftover, one member group of those two should be redundant.

amukkara replied:
We could remove the aow_storage versions. Using thrust::device_vector throughout would be much simpler. Bitvector class is not using any window functionality like say static_set would.

PointKernel replied:
sure, works for me.

amukkara committed:
Remove bulk set operation

amukkara committed:
Remove aow_storage structures

Directly use thrust::device_vector in device-ref code

amukkara committed:
Grow bitvector on device

PointKernel committed:
Move bit_vector to detail namespace

PointKernel committed:
Add missing headers

PointKernel committed:
Clean up type aliases and static constexpr

PointKernel committed:
Add missing headers

PointKernel committed:
Add missing headers

PointKernel committed:
Add allocator template parameter

PointKernel committed:
Clean up docs

PointKernel committed:
Move kernels to a separate file

PointKernel committed:
Make bit_vector_ref a nested type

PointKernel committed:
Remove bv read operator tag

PointKernel committed:
Move implementation details to inl file

PointKernel committed:
Rename bit_vector as dynamic_bitset

PointKernel commented:
/ok to test

PointKernel committed:
Merge remote-tracking branch 'upstream/dev' into trie-bitvector

PointKernel committed:
Cleanups: constexpr instead of inline, TODO, etc

PointKernel commented:
/ok to test

amukkara committed:
Change names of rank, select variables

 rank,   select  ->  rank_true,  select_true

 rank0, select0  ->  rank_false, select_false

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+   *
+   * @return The base rank
+   */
+  __host__ __device__ constexpr uint64_t abs() const noexcept
```
Could we find a more descriptive name for this function and also `set_abs`?

amukkara replied:
How about replacing 'abs' with 'base'?

`rank` struct encodes a list of 4 integers say [1000, 1005, 1006, 1009] as [1000, 5, 6, 9]

what is the standard terminology for such encoding?

PointKernel replied:
Yeah, `base` sounds good.

amukkara replied:
a85b7b70eb22f7c1a03279c4a8674e8c949a6aba

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+   */
+  __host__ __device__ constexpr uint64_t abs() const noexcept
+  {
+    return (static_cast<uint64_t>(abs_hi_) << 8) | abs_lo_;
```
```suggestion
    return (static_cast<uint64_t>(abs_hi_) << CHAR_BIT) | abs_lo_;
```

amukkara replied:
a85b7b70eb22

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+   */
+  __host__ __device__ constexpr void set_abs(uint64_t abs) noexcept
+  {
+    abs_hi_ = static_cast<uint32_t>(abs >> 8);
```
```suggestion
    abs_hi_ = static_cast<uint32_t>(abs >> CHAR_BIT);
```

amukkara replied:
a85b7b70eb22f7c1a03279c4a8674e8c949a6aba

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+  static constexpr size_type bits_per_block  = words_per_block * bits_per_word;  ///< Trivial
+
+  /**
+   * @brief Constructs an empty bitvector
```
```suggestion
   * @brief Constructs an empty biset
```

amukkara replied:
ec6bb1117ed9e7d5489763549a0de818651e6c03

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+
+/**
+ * @brief Bitvector class with rank and select index structures
+ *
```
```suggestion
 *
 * Current limitations:
 * - Stream controls are partially supported due to the use of `thrust::device_vector` as storage
 * - Device ref doesn't support modifiers like `set`, `reset`, etc. 
 *
```

amukkara replied:
ec6bb1117ed9e7d5489763549a0de818651e6c03

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+ * rank and select operation API. It maintains index structures to make both these
+ * new operations close to constant time.
+ * Bitvector construction happens on host, after which the structures are moved to device.
+ * All subsequent read-only operations access device structures only.
```
```suggestion
```
doc to be updated

amukkara replied:
ec6bb1117ed9e7d5489763549a0de818651e6c03

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+};
+
+/**
+ * @brief Bitvector class with rank and select index structures
```
```suggestion
 * @brief Bitset class with rank and select index structures
```

amukkara replied:
ec6bb1117ed9e7d5489763549a0de818651e6c03

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+   *
+   * @param bit Boolean value of new bit to be added
+   */
+  constexpr void append(bool bit) noexcept;
```
```suggestion
  /**
   * @brief Appends the given element `value` to the end of the bitset
   *
   * This API may involve data reallocation if the current storage is exhausted.
   *
   * @param value Boolean value of the new bit to be added
   */
  constexpr void push_back(bool value) noexcept;
```

amukkara replied:
0f1db0fae4cdedfcf7605f996400c647123ea13b

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+   */
+  constexpr void append(bool bit) noexcept;
+
+  /**
+   * @brief Modifies a single bit
+   *
+   * @param index position of bit to be modified
+   * @param bit new value of bit
+   */
+  constexpr void set(size_type index, bool bit) noexcept;
```
```suggestion
  /**
   * @brief Sets the target bit indexed by `index` to a specified `value`.
   *
   * @param index Position of bit to be modified
   * @param value New value of the target bit
   */
  constexpr void set(size_type index, bool value) noexcept;
```

amukkara replied:
0f1db0fae4cdedfcf7605f996400c647123ea13b

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+   */
+  constexpr void set(size_type index, bool bit) noexcept;
+
+  /**
+   * @brief Sets last bit to specified value
+   *
+   * @param bit new value of last bit
+   */
+  constexpr void set_last(bool bit) noexcept;
```
```suggestion
  /**
   * @brief Sets the last bit to a specified value
   *
   * @param value New value of the last bit
   */
  constexpr void set_last(bool value) noexcept;
```

amukkara replied:
0f1db0fae4cdedfcf7605f996400c647123ea13b

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+   */
+  constexpr void set_last(bool bit) noexcept;
+
+  /**
+   * @brief Builds indexes for rank and select
+   */
+  constexpr void build() noexcept;
```
I'm thinking of adding a new data member called
```cuda
  bool is_built_; ///< Flag indicating whether the rank and select indices are built or not
```
We set the flag to `false` in ctor and all modifier functions and at the beginning of query/read functions, we can do something like:
```cuda
  if (not is_built_) {
    this->build();
    is_built_ = true;
  }
```
In this way,  this could be a private function and we can hide the build process from users.

amukkara replied:
This makes sense. We need to do this check within every bulk query operation and ref(). I will take a stab at making the change.

PointKernel replied:
Unless it's needed by trie, let's just focus on bulk operations for now since such a flag in ref APIs would cause host/device synchronization issues that we tend to ignore for now.

amukkara replied:
a629730842d08d41696b

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+  /**
+   * @brief Bulk get operation
+   *
+   * @tparam KeyIt Device-accessible iterator to keys
+   * @tparam OutputIt Device-accessible iterator to outputs
```
```suggestion
   * @tparam KeyIt Device-accessible iterator to keys whose `value_type` can be converted to bitset's `size_type`
   * @tparam OutputIt Device-accessible iterator to outputs whose `value_type` can be constructed from boolean type
```

amukkara replied:
81b8e907e0acc19a5eabfaeeb927dd39db103739

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+  constexpr void build() noexcept;
+
+  /**
+   * @brief Bulk get operation
```
```suggestion
   * @brief For any element `keys_begin[i]` in the range `[keys_begin, keys_end)`, stores the boolean value at position `keys_begin[i]` to `output_begin[i]`.
```

amukkara replied:
b2c88ded9df8bfbd45a98be9f88c3c9e51bcaffc

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+   * @param stream Stream to execute get kernel
+   */
+  template <typename KeyIt, typename OutputIt>
+  constexpr void get(KeyIt keys_begin,
```
```suggestion
  constexpr void test(KeyIt keys_begin,
```
`test` instead of `get` to align with STL/boost.

amukkara replied:
f96468df7b03a5ad8dd40eba9463f74fceb09142

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+  /**
+   * @brief Bulk rank operation
+   *
+   * @tparam KeyIt Device-accessible iterator to keys
+   * @tparam OutputIt Device-accessible iterator to output ranks
```
please be explicit about the iterator's `value_type`

amukkara replied:
81b8e907e0acc19a5eabfaeeb927dd39db103739

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+  /**
+   * @brief Bulk select operation
+   *
+   * @tparam KeyIt Device-accessible iterator to keys
+   * @tparam OutputIt Device-accessible iterator to outputs
```
specify iterator's `value_type`

amukkara replied:
81b8e907e0acc19a5eabfaeeb927dd39db103739

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+     *
+     * @return Index of next set bit
+     */
+    [[nodiscard]] __device__ size_type find_next_set(size_type key) const noexcept;
```
```suggestion
    [[nodiscard]] __device__ size_type find_next(size_type key) const noexcept;
```
https://www.boost.org/doc/libs/1_83_0/libs/dynamic_bitset/dynamic_bitset.html#find_next

amukkara replied:
f96468df7b03a5ad8dd40eba9463f74fceb09142

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+     *
+     * @return Value of bit at position specified by key
+     */
+    [[nodiscard]] __device__ constexpr bool get(size_type key) const noexcept;
```
```suggestion
    [[nodiscard]] __device__ constexpr bool test(size_type key) const noexcept;
```

amukkara replied:
f96468df7b03a5ad8dd40eba9463f74fceb09142

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+     *
+     * @return Word at position specified by index
+     */
+    [[nodiscard]] __device__ constexpr slot_type get_word(size_type word_id) const noexcept;
```
```suggestion
    [[nodiscard]] __device__ constexpr slot_type word(size_type word_id) const noexcept;
```

amukkara replied:
59457e91c5eba5f93d45563c54336a9282f26b2b

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+class dynamic_bitset {
+ public:
+  using size_type = std::size_t;  ///< size type to specify bit index
+  using slot_type = uint64_t;     ///< Slot type
```
Question: Should this be just `word_type`?

amukkara replied:
Yes, slot_type is a vestige from aow_storage usage

amukkara replied:
63ed55224354aacccde9054fc3ab698e91186b22

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+   * @param stream Stream to execute ranks kernel
+   */
+  template <typename KeyIt, typename OutputIt>
+  constexpr void ranks(KeyIt keys_begin,
```
```suggestion
  constexpr void rank(KeyIt keys_begin,
```

amukkara replied:
c753f1082bf34b074023d6c80fe1e6ad59d6d075

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+   * @param stream Stream to execute selects kernel
+   */
+  template <typename KeyIt, typename OutputIt>
+  constexpr void selects(KeyIt keys_begin,
```
```suggestion
  constexpr void select(KeyIt keys_begin,
```

amukkara replied:
c753f1082bf34b074023d6c80fe1e6ad59d6d075

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+     *
+     * @return Position of Nth not-set bit
+     */
+    [[nodiscard]] __device__ constexpr size_type select0(size_type count) const noexcept;
```
```suggestion
    [[nodiscard]] __device__ constexpr size_type select_false(size_type count) const noexcept;
```
`select_false` instead of `select0` based on the offline discussion

amukkara replied:
c7faed0992d792f37919b8749949d0d3e7cf88ee

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,359 @@
+     * @return index in ranks which corresponds to highest rank less than count (least upper bound)
+     */
+    template <typename SelectsRef, typename RanksRef>
+    [[nodiscard]] __device__ constexpr size_type get_initial_rank_estimate(
```
```suggestion
    [[nodiscard]] __device__ constexpr size_type initial_rank_estimate(
```
`get_` prefixes rarely provide useful information so let's just remove it.

amukkara replied:
59457e91c5eba5f93d45563c54336a9282f26b2b

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/kernels.cuh:
```diff
@@ -0,0 +1,234 @@
+                                     ValueIt outputs,
+                                     size_type num_keys)
+{
+  uint32_t const loop_stride = gridDim.x * blockDim.x;
+  uint32_t key_id            = blockDim.x * blockIdx.x + threadIdx.x;
```
```suggestion
  cuco::detail::index_type const loop_stride = gridDim.x * blockDim.x;
  cuco::detail::index_type key_id            = blockDim.x * blockIdx.x + threadIdx.x;
```
Similar to
https://github.com/NVIDIA/cuCollections/blob/c0cce866368396ced18a37a0d30287729d6af6c4/include/cuco/detail/common_kernels.cuh#L74-L75, please use `cuco::detail::index_type` consistently to avoid potential overflow issue. The same to other kernels.

amukkara replied:
b26f326b6929531b6517305031e8c816402fd05a

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/kernels.cuh:
```diff
@@ -0,0 +1,234 @@
+ * @param outputs Begin iterator to outputs
+ * @param num_keys Number of input keys
+ */
+template <typename BitvectorRef, typename KeyIt, typename ValueIt, typename size_type>
```
```suggestion
template <typename BitsetRef, typename KeyIt, typename ValueIt, typename size_type>
```

amukkara replied:
ec6bb1117ed9e7d5489763549a0de818651e6c03

amukkara committed:
Rename members and methods of `rank`

abs  -> base
rels -> offsets

amukkara committed:
Rename bitvector to bitset

amukkara committed:
Remove `get_` prefixes in method names

amukkara committed:
Use rank_type

This will allow us to rename ranks() to rank()

amukkara committed:
Rename bulk API methods

ranks -> rank
selects -> select

amukkara committed:
Use cuco::detail::index_type in kernels

amukkara committed:
Change some API to match boost dynamic_bitset

amukkara committed:
Rename file to match previous API change

amukkara committed:
More API changes

amukkara committed:
Rename slot_type to word_type

amukkara committed:
Specify iterator's value_type in doxygen comments

amukkara committed:
Comments

PointKernel commented on include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh:
```diff
@@ -0,0 +1,377 @@
+   *
+   * @return grid size
+   */
+  // TODO: to be moved to the CUDA utility header
+  constexpr size_type default_grid_size(size_type num_elements) const noexcept;
```
To be removed once #359 gets merged.

PointKernel committed:
Merge remote-tracking branch 'upstream/dev' into trie-bitvector

PointKernel committed:
Use detail CUDA utilities to determine grid size

amukkara committed:
Make build() a private member

Query methods will check and perform build()

amukkara committed:
Merge branch 'trie-bitvector' of github.com:amukkara/cuCollections into trie-bitvector

PointKernel commented:
/ok to test

PointKernel committed:
Minor doc updates

PointKernel committed:
Clean up kernels with cuda utilities

PointKernel committed:
Minor style cleanup

PointKernel committed:
Cleanups: renaming + update docs

PointKernel committed:
Consistently use the same allocator for intermediate vars

PointKernel committed:
Make build process exposed to CUDA stream

PointKernel committed:
Cleanups + deallocate before return

PointKernel commented:
/ok to test

sleeepyjack reviewed (approved):
LGTM
